src/rules/add-claim-based-on-relationships.js

```diff
@@ -0,0 +1,121 @@
+/**
+* @title Adds a claim based on relationships
+* @overview Adds a claim based on the relationships the subject has in Scaled Access
+* @gallery true
+* @category marketplace
+*
+* This rule adds a claim to the access token based on relationships the subject has in Scaled Access.
+* 
+* Example of a resulting claim:
+* ```
+* "https://example.com/relationships": [
+*    {
+*      "relationshipType": "is_admin_of",
+*      "to": {
+*        "id": "c7b134f9-28a2-4dcc-b345-affa18977ddf",
+*        "type": "subscription"
+*      }
+*    },
+*    {
+*      "relationshipType": "is_sports_member_of",
+*      "to": {
+*        "id": "affa18977ddf-4dcc-b345-c7b134f9-28a2",
+*        "type": "subscription"
+*      }
+*    }
+* ]
+* ```
+* This is done through an API call to Scaled Access' Relationship Management API using a machine-to-machine token.
+* More info can be found at https://docs.scaledaccess.com/?path=integration-with-auth0
+* 
+* A number of rule settings are required:
+* - GROUPS4AUTH0_CLIENTID: The Client ID of the machine-to-machine application.
+* - GROUPS4AUTH0_CLIENTSECRET: The Client secret of the machine-to-machine application.
+* - GROUPS4AUTH0_BASEURL: The base URL for the Relationship Management API, e.g. https://api.int.scaledaccess.com/privategroups-v2.
+* - GROUPS4AUTH0_TENANT: Your Scaled Access tenant code.
+* - GROUPS4AUTH0_CUSTOMCLAIM: A namespaced custom claim name of your choice. The name must be a URL.
+* - GROUPS4AUTH0_AUDIENCE: The audience in the machine-to-machine token.
+*/
+function addClaimBasedOnRelationships(user, context, callback) {
+    const fetch = require("node-fetch");
+    const { URLSearchParams } = require('url');
+    const jwt = require('jsonwebtoken');
+
+    const getM2MToken = () => {
+        if (global.m2mToken && global.m2mTokenExpiryInMillis > new Date().getTime() + 60000) {
```

Comment:
We're caching the m2m token. Is there a better way?

Reply:
Just a `let m2mToken` at the top-level scope should suffice. Anything saved here will not persist across requests anyways.

Reply:
Are you saying there's no way to cache the m2m token? For this to work, we need something that can persist across requests.

From https://auth0.com/docs/best-practices/rules-best-practices/rules-environment-best-practices:
> The global object can also be used to cache expensive resources, such as an Access Token for a third-party (e.g., logging) API that provides non user-specific functionality or an Access Token to your own API defined in Auth0 and obtained by using Client Credentials flow.

Reply:
🤔  Let me check with the team on that. I haven't seen it used that way before but the docs make it seem like that would work.

Reply:
I think this will work as you have it. The global persists anywhere from ~0 to ~20 minutes (container lifetime) so storing it like you have will help with rate limits and performance.